In [1]:
import pandas as pd
import numpy as np
import json

In [2]:
df_real = pd.read_csv('../Cleaned Data/Real_News.csv')
df_real = df_real[['url','title','text']]
print(df_real.shape)
df_real.head()



(905, 3)


url  \
0  https://www.cnn.com/2020/03/29/politics/trump-...   
1  https://www.cnn.com/world/live-news/coronaviru...   
2  https://www.cnn.com/2020/03/29/politics/trump-...   
3  https://www.cnn.com/2020/03/29/opinions/donald...   
4  https://www.cnn.com/2020/03/29/politics/us-ele...   

                                               title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                         Donald Trump's special day   
4             How long will Trump's poll surge last?   

                                                text  
0  Washington (CNN) President Donald Trump said S...  
1  The exterior of Elmhurst Hospital in Queens on...  
2  (CNN) President Donald Trump acknowledged Sund...  
3  Sign up to get our new weekly column as a news...  
4  (CNN) The 2020 race, publicly at least, has la...

In [3]:
df_real['label'] = 1
df_real.dropna(inplace=True)
print(df_real.shape)
df_real.drop(['url'],axis=1,inplace=True)
df_real.head()

(905, 4)


title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                         Donald Trump's special day   
4             How long will Trump's poll surge last?   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  (CNN) President Donald Trump acknowledged Sund...      1  
3  Sign up to get our new weekly column as a news...      1  
4  (CNN) The 2020 race, publicly at least, has la...      1

In [4]:
df_fake = pd.read_csv('../Cleaned Data/Human_Fake_News.csv')

In [5]:
print(df_fake.shape)
df_fake.head()

(12273, 23)


Unnamed: 0                                      uuid  ord_in_thread  \
0           0  6a175f46bcd24d39b3e962ad0f29936721db70db              0   
1           1  2bdc29d12605ef9cf3f09f9875040a7113be5d5b              0   
2           2  c70e149fdd53de5e61c29281100b9de0ed268bc3              0   
3           3  7cf7c15731ac2a116dd7f629bd57ea468ed70284              0   
4           4  0206b54719c7e241ffe0ad4315b808290dbe6c0f              0   

                 author                      published  \
0     Barracuda Brigade  2016-10-26T21:41:00.000+03:00   
1  reasoning with facts  2016-10-29T08:47:11.259+03:00   
2     Barracuda Brigade  2016-10-31T01:41:49.479+02:00   
3                Fed Up  2016-11-01T05:22:00.000+02:00   
4                Fed Up  2016-11-01T21:56:00.000+02:00   

                                               title  \
0  Muslims BUSTED: They Stole Millions In Gov’t B...   
1  Re: Why Did Attorney General Loretta Lynch Ple...   
2  BREAKING: Weiner Cooperating With FBI On Hilla...   
3  PIN DROP SPEECH BY FATHER OF DAUGHTER Kidnappe...   
4  FANTASTIC! TRUMP'S 7 POINT PLAN To Reform Heal...   

                                                text language  \
0  Print They should pay all the back all the mon...  english   
1  Why Did Attorney General Loretta Lynch Plead T...  english   
2  Red State : \nFox News Sunday reported this mo...  english   
3  Email Kayla Mueller was a prisoner and torture...  english   
4  Email HEALTHCARE REFORM TO MAKE AMERICA GREAT ...  english   

                         crawled             site_url  ... spam_score  \
0  2016-10-27T01:49:27.168+03:00  100percentfedup.com  ...      0.000   
1  2016-10-29T08:47:11.259+03:00  100percentfedup.com  ...      0.000   
2  2016-10-31T01:41:49.479+02:00  100percentfedup.com  ...      0.000   
3  2016-11-01T15:46:26.304+02:00  100percentfedup.com  ...      0.068   
4  2016-11-01T23:59:42.266+02:00  100percentfedup.com  ...      0.865   

                                        main_img_url replies_count  \
0  http://bb4sp.com/wp-content/uploads/2016/10/Fu...             0   
1  http://bb4sp.com/wp-content/uploads/2016/10/Fu...             0   
2  http://bb4sp.com/wp-content/uploads/2016/10/Fu...             0   
3  http://100percentfedup.com/wp-content/uploads/...             0   
4  http://100percentfedup.com/wp-content/uploads/...             0   

   participants_count likes  comments  shares  type  \
0                   1     0         0       0  bias   
1                   1     0         0       0  bias   
2                   1     0         0       0  bias   
3                   0     0         0       0  bias   
4                   0     0         0       0  bias   

                                          clean_text  \
0  print pay back money plus interest entire fami...   
1  attorney general loretta lynch plead fifth bar...   
2  red state  fox news sunday reported morning an...   
3  email kayla mueller prisoner tortured isis cha...   
4  email healthcare reform make america great sin...   

                                         clean_title  
0        muslims busted stole millions govt benefits  
1         attorney general loretta lynch plead fifth  
2  breaking weiner cooperating fbi hillary email ...  
3  pin drop speech father daughter kidnapped kill...  
4  fantastic trumps  point plan reform healthcare...  

[5 rows x 23 columns]

In [6]:
df_fake = df_fake[['site_url','title','text']]
df_fake.dropna(inplace=True)
df_fake.rename(columns={"site_url": "url"}, inplace=True)
chosen_idx = np.random.choice(df_fake.shape[0], replace=False, size=df_real.shape[0])
df_fake_trim = df_fake.iloc[chosen_idx]

In [7]:
df_fake_trim['label'] = 0
print(df_fake_trim.shape)
df_fake_trim.drop(['url'], axis=1, inplace= True)
df_fake_trim.head()


(905, 4)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


title  \
11375  Syrian War Report – November 2, 2016: ISIS and...   
12039          World wildlife 'falls by 58% in 40 years'   
1068         Black Agenda Radio for Week of Nov 21, 2016   
271    The U.S./Turkey Plan For “Seizing, Holding, An...   
6410   Furious Eric Holder Just Issued A Dire Warning...   

                                                    text  label  
11375  Trump Whistles His Dogs ‹ › South Front Analys...      0  
12039  Donations World wildlife 'falls by 58% in 40 y...      0  
1068   News, information and analysis from the black ...      0  
271    By Brandon Turbeville As the U.S. Presidential...      0  
6410   Comments \nFormer US Attorney General Eric Hol...      0

In [8]:
df_grover = pd.read_csv('../Cleaned Data/Neural_Fake_News.csv')
df_grover = df_grover[['title','text']]
print(df_grover.shape)
df_grover.dropna(inplace=True)
print(df_grover.shape)
df_grover.head()
df_grover['label'] = 2
df_grover.head()

(908, 2)
(908, 2)


title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                  Tracking Covid-19 cases in the US   
4                         Donald Trump's special day   

                                                text  label  
0  President Donald Trump, on his first official ...      2  
1  By Dylan Welch\nApril 1, 2015. (Bloomberg) -- ...      2  
2  Donald Trump won his home state of California ...      2  
3  Editor’s note: Readers and visitors in the New...      2  
4  The wish of many in the press was for the morn...      2

# Experiment Number 1
## Naive Bayes, Real Human News and Fake Human News
905 instances of each 

Class: Label

Real: 0

Fake: 1

In [9]:
#combine both the real and trimmed fake news into one df

# combined = pd.concat([df_real,df_fake_trim,df_grover],ignore_index=True) # For 3 class classification 
# combined = pd.concat([df_real,df_grover],ignore_index=True) # For real vs neural fake
combined = pd.concat([df_real,df_fake_trim],ignore_index=True) # For real vs human fake

In [10]:
print(combined.head())
combined.tail()

                                               title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                         Donald Trump's special day   
4             How long will Trump's poll surge last?   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  (CNN) President Donald Trump acknowledged Sund...      1  
3  Sign up to get our new weekly column as a news...      1  
4  (CNN) The 2020 race, publicly at least, has la...      1  


title  \
1805  Mike Pence Drapes Shawl Over Immodest Lady Jus...   
1806        Comment on All Saints Day by Steven Broiles   
1807  Kenyan marathon runner Rita Jeptoo banned 2 mo...   
1808  8 Things You Need To Know About The Dakota Acc...   
1809  Fans Asked George Takei To Tweet Something Pos...   

                                                   text  label  
1805  Trump Raises Concern Over Members Of Urban Com...      0  
1806  Posted on November 1, 2016 by joandarc | 8 Com...      0  
1807  News Bulletin Rita Jeptoo celebrates with the ...      0  
1808  North Dakota is ablaze as a result of proteste...      0  
1809  Comments \nLegendary Star Trek actor turned LG...      0

In [11]:
import nltk
from wordcloud import WordCloud
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, accuracy_score, recall_score, precision_score

In [12]:
combined_text = combined['text'].values
X_title =combined['title'].values
labels = combined['label'].values
combined.groupby('label').agg('count')

title  text
label             
0        905   905
1        905   905

In [13]:
tfidf = TfidfVectorizer(ngram_range=(1,2), max_df= 0.85, min_df= 0.01)

In [14]:
combined_text = tfidf.fit_transform(combined_text)
# combined['title'] = tfidf.fit_transform (combined['title'])

In [15]:
print("tf idf vectorized text shape: {}".format(combined_text.shape))
print("original text shape: {}".format(combined['text'].shape))


tf idf vectorized text shape: (1810, 10200)
original text shape: (1810,)


In [16]:
train_text, test_text, train_labels, test_labels  = train_test_split(combined_text, labels, test_size=0.2)


In [17]:
import collections

print("size of train_text: {}".format(train_text.shape))

print("size of train_labels: {}".format(train_labels.shape))
collections.Counter(train_labels)


size of train_text: (1448, 10200)
size of train_labels: (1448,)


Counter({0: 732, 1: 716})

In [18]:
print("size of test_text: {}".format(test_text.shape))

print("size of test_labels: {}".format(test_labels.shape))
collections.Counter(test_labels)



size of test_text: (362, 10200)
size of test_labels: (362,)


Counter({0: 173, 1: 189})

In [19]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()


In [20]:
nb.fit(train_text,train_labels)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [21]:
train_pred = nb.predict(train_text)

In [22]:
print('Naive Bayes In Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(train_labels, train_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(train_labels, train_pred)*100))

Naive Bayes In Sample F1 and Accuracy Scores:
F1 score 97.1%
Accuracy score 97.1%


In [23]:
combined.loc[np.where(train_labels != train_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
0         20    20
1         22    22

In [24]:
test_pred = nb.predict(test_text)
print('Naive Bayes Out of Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(test_labels, test_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(test_labels, test_pred)*100))

Naive Bayes Out of Sample F1 and Accuracy Scores:
F1 score 94.75%
Accuracy score 94.75%


In [25]:
combined.loc[np.where(test_labels != test_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
1         19    19

# Experiment 2
## Augmenting with Neural Fake News

905 Real News, 905 Fake (neural and human combined)

80% Train = 1448
20% Test = 362 (1/2 Real, 1/2 Human Fake)

Train on (1/2 real news, 1/2 fake news (1/4 human fake + 1/4 neural fake))

Human Fake = 543 (362 for train, 181 for test)

Neural Fake = 362 for train

In [26]:
print(df_grover.shape)
print(df_real.shape)
print(df_fake.shape)

(908, 3)
(905, 3)
(12273, 3)


In [27]:
trim_size = int(df_real.shape[0] * 0.6)
chosen_idx = np.random.choice(df_fake.shape[0], replace=False, size=trim_size)
df_fake_trim_2 = df_fake.iloc[chosen_idx]
df_fake_trim_2['label'] = 0
print(df_fake_trim_2.shape)
df_fake_trim_2.drop(['url'], axis=1, inplace= True)
df_fake_trim_2.head()

(543, 4)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


title  \
8208  Secretary of State? Bolton Advises Trump to “C...   
6822  The Top Democrat In The Senate Just Accused FB...   
4651  Here’s what happened when a Hillary supporting...   
631   Ashton Kutcher Rescues 6,000 Sex Trafficking V...   
1730    Michael Moore Wants A Spot in Hillary’s Cabinet   

                                                   text  label  
8208  \nThis article was written by Kurt Nimmo and o...      0  
6822  Here is the full text of Sen. Reid’s letter to...      0  
4651  VIDEOS Here’s what happened when a Hillary sup...      0  
631   Christopher Ashton Kutcher is a well-known fig...      0  
1730  Given the shameless, over-the-top, almost brea...      0

In [28]:
trim_size = int(df_real.shape[0]*0.4)
chosen_idx = np.random.choice(df_grover.shape[0], replace=False, size=trim_size)
df_grover_trim = df_grover.iloc[chosen_idx]
df_grover_trim['label'] = 0
print(df_grover_trim.shape)
df_grover_trim.head()

(362, 3)


/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


title  \
242  ‘The Other Option Is Death’: New York Starts S...   
202  Facebook, Google and Twitter Struggle to Handl...   
892  ‘A mess in America’: Why Asia now looks safer ...   
589        Coronavirus upends the battle for the House   
627  Here are remote jobs that could keep you worki...   

                                                  text  label  
242  If you were searching for a new way to disable...      0  
202  Invisible ballots and ballot-in-boxes may slow...      0  
892  WASHINGTON — If you listen to President Donald...      0  
589  CORONAVIA, Md. (April 1, 2020) - The rising ra...      0  
627  If these job offers fail, we may be lost to a ...      0

In [29]:
print(df_real.shape)
df_real.head()

(905, 3)


title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                         Donald Trump's special day   
4             How long will Trump's poll surge last?   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  (CNN) President Donald Trump acknowledged Sund...      1  
3  Sign up to get our new weekly column as a news...      1  
4  (CNN) The 2020 race, publicly at least, has la...      1

In [30]:
#Create Test Set first from trimmed
trim_size = int(df_real.shape[0]*.2)
mask =  np.random.choice(df_real.shape[0], trim_size, replace=False) 
neg_mask = [i for i in range(0, df_real.shape[0]) if i not in mask]
df_real_test = df_real.iloc[mask]
df_real_train = df_real.iloc[neg_mask]
print(df_real_test.shape)
print(df_real_train.shape)

(181, 3)
(724, 3)


In [31]:
#Create Test Set first from trimmed
print(df_fake_trim_2.shape)
trim_size = int(df_fake_trim_2.shape[0]*(1/3))
mask =  np.random.choice(df_fake_trim_2.shape[0], trim_size, replace=False) 
neg_mask = [i for i in range(0, df_fake_trim_2.shape[0]) if i not in mask]
df_fake_trim_2_test = df_fake_trim_2.iloc[mask]
df_fake_trim_2_train = df_fake_trim_2.iloc[neg_mask]
print(df_fake_trim_2_test.shape)
print(df_fake_trim_2_train.shape)

(543, 3)
(181, 3)
(362, 3)


In [32]:
#combine both the real and trimmed fake news into one df
combined_train = pd.concat([df_real_train,df_fake_trim_2_train,df_grover_trim],ignore_index=True)
combined_test = pd.concat([df_real_test,df_fake_trim_2_test],ignore_index=True)

In [33]:
print(combined_train.shape)
print(combined_train.head())
combined_train.tail()

(1448, 3)
                                               title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2                         Donald Trump's special day   
3  A breakdown of false and misleading statements...   
4  New York coronavirus deaths soar past 960 as N...   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  Sign up to get our new weekly column as a news...      1  
3  (CNN) President Donald Trump has made numerous...      1  
4  (CNN) The mammoth fight against coronavirus in...      1  


title  \
1443  Coronavirus Diverts U.S. Aircraft Carrier From...   
1444  Pennsylvania hit by more jobless claims than a...   
1445  Space Force leaders' quest: A gender-neutral f...   
1446  What's Inside The Senate's $2 Trillion Coronav...   
1447  On Job Just 6 Weeks, U.K.’s Finance Chief Shin...   

                                                   text  label  
1443  After making its first flight in Asia in Febru...      0  
1444  By Marc Levy (Reuters) - Pennsylvania could be...      0  
1445  Amid the last dozen or so directives from Stra...      0  
1446  Posted Tuesday, April 1, 2020 9:25 am\nRep. Fr...      0  
1447  UK financial secretary Jack Straw on Thursday ...      0

In [34]:
print(combined_test.shape)
print(combined_test.head())
combined_test.tail()

(362, 3)
                                               title  \
0  Gilead Revamps Process for Emergency Requests ...   
1                Coronavirus: Live BBC News coverage   
2   Lou Ferrigno gives it another go in Santa Monica   
3  For France, Coronavirus Tests a Vaunted Health...   
4  Shareholder suit accuses Sen. Richard Burr of ...   

                                                text  label  
0  Gilead Sciences Inc. will stop considering ind...      1  
1  Video\n\nLive BBC News coverage of the coronav...      1  
2  Bodybuilder turned actor Lou Ferrigno has put ...      1  
3  PARIS — One of the world’s best health care sy...      1  
4  Burr was also in the hot seat after NPR report...      1  


title  \
357  The Yellow Brick Road to Extinction - Russia N...   
358  Kunst am Werk: Wie UralMasch zur Bilder-Galeri...   
359  Actor Mark Ruffalo Joins Activists In ND To Pr...   
360  Is Russia Hiding The Existence Of An Advanced ...   
361      World's Oldest Person Had Smoked For 76 Years   

                                                  text  label  
357  This post was originally published on this sit...      0  
358  Audio-Slideshows Kunst am Werk: Wie UralMasch ...      0  
359  Mark Ruffalo is one the movie industry’s most ...      0  
360  Peculiar supernatural activity which is suspec...      0  
361  Report Copyright Violation World's Oldest Pers...      0

In [35]:
train_text = combined_train['text'].values
test_text = combined_test['text'].values
train_labels = combined_train['label'].values
test_labels = combined_test['label'].values

In [36]:
type(train_text)
print(train_text.shape)
print(test_text.shape)
fit_text = np.concatenate((train_text, test_text))
print(fit_text.shape)

(1448,)
(362,)
(1810,)


In [37]:
tfidf_2 = TfidfVectorizer(ngram_range=(1,2), max_df= 0.85, min_df= 0.01)
tfidf_2.fit(fit_text)
train_text = tfidf_2.transform(train_text)
print(type(train_text))
print("tf idf vectorized text shape: {}".format(train_text.shape))
print("original text shape: {}".format(combined_train['text'].shape))

test_text = tfidf_2.transform(test_text)
print(type(test_text))
print("tf idf vectorized text shape: {}".format(test_text.shape))
print("original text shape: {}".format(combined_test['text'].shape))

<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (1448, 9935)
original text shape: (1448,)
<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (362, 9935)
original text shape: (362,)


In [38]:
nb_2 = MultinomialNB()
nb_2.fit(train_text,train_labels)
train_pred = nb_2.predict(train_text)

In [39]:
print('Naive Bayes In Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(train_labels, train_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(train_labels, train_pred)*100))

Naive Bayes In Sample F1 and Accuracy Scores:
F1 score 91.99%
Accuracy score 91.99%


In [40]:
combined.loc[np.where(train_labels != train_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
0         66    66
1         50    50

In [41]:
test_pred = nb_2.predict(test_text)
print('Naive Bayes Out of Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(test_labels, test_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(test_labels, test_pred)*100))

Naive Bayes Out of Sample F1 and Accuracy Scores:
F1 score 93.07%
Accuracy score 93.09%


In [42]:
combined.loc[np.where(test_labels != test_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
1         25    25